pkg/tls/tls_test.go

```diff
@@ -0,0 +1,241 @@
+		fmt.Errorf("unable to load last known good cert from disk %s", err)
+	}
+	block, _ := pem.Decode(knownGoodBlob)
+	if block == nil || block.Type != "CERTIFICATE" {
+		fmt.Errorf("failed to decode PEM block containing public key")
+	}
+
+	knownGoodCerts, err := x509.ParseCertificates(block.Bytes)
+	if err != nil {
+		fmt.Errorf("unable to parse certificates %s", err)
+	}
+	goodCert := knownGoodCerts[0]
+	return goodCert
+}
+
+const cn = "dummy-test-certificate.local"
+const knownGoodLoc = "./testdata/known_good_cert.pem"
+const knownGoodLocCA = "./testdata/known_good_certCA.pem"
+const updateSSLVar = "UPDATE_KNOWN_SSL_CERT"
+
+var orgs = []string{cn}
+
+func TestNewPrivateKey(t *testing.T) {
+	newPK, err := NewPrivateKey()
+	if err != nil {
+		t.Errorf("failed to create private key")
+	}
+	if newPK.Validate() != nil {
+		t.Errorf("Failed to validate private key")
+	}
+	// the exponent which is effectively the length is encoded inside PublicKey
+	keyLength := newPK.PublicKey.N.BitLen()
+	if keyLength < 2048 {
+		t.Errorf("insecure key length detected: %d", keyLength)
+	}
+}
+
+func TestNewCA(t *testing.T) {
+	pk, cert, err := NewCA(cn)
+	if err != nil {
+		t.Errorf("failed to create new CA %s", err)
+	}
+	if pk.Validate() != nil {
+		t.Errorf("failed to validate RSA Private key")
+	}
+	if os.Getenv(updateSSLVar) == "true" {
+		// reuse functionality from marshal.go
+		pemEncodedCert, err := CertAsBytes(createKnownGoodCert(true))
+		if err != nil {
+			t.Errorf("Cannot PEM encode certificate")
+		}
+		err = ioutil.WriteFile(knownGoodLocCA, pemEncodedCert, 0644)
+		if err != nil {
+			t.Errorf("unable to save PEM cert %s", err)
+		}
+	}
+	goodCert := loadGoodCert(knownGoodLocCA)
+	cert.SerialNumber = big.NewInt(0)
+	cert.NotBefore = time.Unix(0, 0)
+	cert.NotAfter = time.Unix(0, 0)
+	// zero out placeholders
+	var zeroBytes []byte
+	var noExt []pkix.Extension
+	var noNames pkix.Name
+	goodCert.Raw = zeroBytes
+	cert.Raw = zeroBytes
+	cert.RawTBSCertificate = zeroBytes
+	goodCert.RawTBSCertificate = zeroBytes
+	cert.RawSubjectPublicKeyInfo = zeroBytes
+	goodCert.RawSubjectPublicKeyInfo = zeroBytes
+	cert.RawSubject = zeroBytes
+	goodCert.RawSubject = zeroBytes
+	// TODO issuers should be the same?
+	goodCert.RawIssuer = zeroBytes
+	cert.RawIssuer = zeroBytes
+	// since we have changed the fields, signature is now invalid
+	cert.Signature = zeroBytes
+	goodCert.Signature = zeroBytes
+	// TODO: do I really need to muck with asn1?
+	cert.Extensions = noExt
+	// our CA is minimal as of now, so let's zero out judiciously
+	goodCert.Issuer = noNames
+	goodCert.Extensions = noExt
+	goodCert.Issuer = noNames
+	goodCert.Subject = noNames
+	goodCert.KeyUsage = 37
+	goodCert.DNSNames = nil
+	goodCert.IPAddresses = nil
+	goodCert.ExtKeyUsage = nil
+	goodCert.Issuer.CommonName = cn
+	goodCert.Subject.Names = nil
+	cert.Issuer.Names = nil
+	cert.Subject.CommonName = ""
+	cert.Subject.Names = nil
+	if diff := deep.Equal(cert, goodCert); diff != nil {
```

Comment:
MUST use reflect.DeepEqual() for the test - unfortunately deep.Equal sometimes hides differences